### ISSUE TITLE
Add initial code from the PSyclone repository

### ISSUE THREAD

**Issue by rupertford** [stfc/fgenerator#1]
The Python code generation software has been developed as part of the PSyclone repository, currently hosted by the Met Office SRS. This ticket will take the latest version from the PSyclone repository and add it here.

**Comment by rupertford** [comment]
Commit a8e33d1 adds in the existing code into this repository. This was previously stored in two files, one with the code and the other with tests. I have split the code into gen, base, modify and subclass.

gen contains all of the classes used to create fortran code the parent class ProgUnitGen and the function bubble_up_type. The latter two pieces of code are only used by the classes within gen and also require some of the gen classes so keeping them together avoids circular dependencies.

base contains the basegen class used by all classes in gen and a utility function that is only used by basegen.

modify contains all functions used to modify existing fortran source that has been parsed by fparser. At the moment there is only the adduse function as that is all that we have required until now.

subclass contains the OpenMP Directive class which is used in conjunction with the gen classes and specializes the fparser comment class allowing us to reason about directives.

I have also modified the tests so that they all pass or xfail.
